### ISSUE TITLE
Migrating issues from Redmine to GitHub

### ISSUE THREAD

**Issue by jgrocha** [qgis/QGIS-Enhancement-Proposals#141]
# QGIS Enhancement: Moving issues from Redmine to GitHub

**Date** 2019/03/07

**Author** Jorge Gustavo (@jgorcha), Régis Haubourg
**Contact** jgr@geomaster.pt,  regis.haubourg at gmail.com

**Version** QGIS (all)

# Summary
QGIS issues are being tracked in Redmine. Our Redmine is becoming obsolete, difficult to upgrade and difficult to use by newcomers. Managing the issues in the same repository as the code (currently in GitHub), has technical advantages and "community" related advantages as well.

This proposal aims to be as specific as possible. The goal is that it can be tested and verified before deployment by any developer.

## Proposed Solution

The proposed solution is to migrate all issues from Redmine to GitHub. 

The Redmine instance can remain online for a while. But if we decide in the future to move to GitLab, we just need to migrate GitHub (because all the issue history is already in GitHub).

#### Overall workflow

1. Proposal for new [QGIS GitHub labels](#qgis-github-labels)
1.1 Define how to map Redmine attributes to GitHub labels.
1.1 Create the labels in QGIS/qgis. Proposal: Bug, Feature request, Feedback, Priority: low, Priority: normal, Priority: high, Regression
1. Create a map between Redmine users and GitHub account
1.1 Get a list of current user names (ie first name, last name) from Redmine (query done)
1.1 Create a very simple form on Redmine site, that receives the user name as parameter and asks for the GitHub account (code done)
1.1 Send an email (from redmine@qgis.org) to each user with the link for the form to provide his GitHub username (script to the written)
1. Create [issue templates](https://help.github.com/en/articles/creating-issue-templates-for-your-repository) for bugs and feature requests (Done https://github.com/qgis/QGIS/pull/9438)
1. Contact GitHub crew to allow more request per hour. Right now, even with 1 sec. delay between each request (to create issue or comment), the script fails. _Our bot has been blocked!_
1. Run the migration (Migration workflow detailed below)
1.1 Check for errors
1.1.1 Delete the last issue migrated and rerun the script from it
1.1 Check if all issues and comments were created
1. Update all Redmine issues. Add a comment to each saying that the issue was migrated do GitHub and add a link to the new issue on GitHub.
1. Put Redmine read only
1. Start mirroring QGIS repository on GitLab.
1. Announce that issues were migrated and from now on the GitHub should be used to fill issues
1.1 Update QGIS documentation
1.2 Use all channels to spread the news

#### Migration workflow

![image](https://user-images.githubusercontent.com/163681/53957561-8cbebc00-40d6-11e9-8147-344707cfe47d.png)

1. Get issues from [Redmine to GitHub](https://gitlab.com/jgrocha/redmine2github). All issues will be transferred to QGIS code repository (some are related with the QGIS Plugin site and some other ones, but only a few).
1. Create a mapping between Redmine attributes and GitHub labels
1. Create a mapping between Redmine users and GitHub users
4. Run migration with: Issues x label map x account map. The result will be new issues on GitHub, plus a key pair json with: redmine id → GitHub id
4.1 Reformat Textile to Markdown
4.2 Assign issue to original assignee if we have the usermap
5. Run migration again with: issue map. This second step is necessary to replace all occurrences of Redmine with GitHub ids.

### QGIS GitHub labels

[Labels in GitHub](https://help.github.com/en/articles/about-labels) are used for PR and for issues within the repository. 

We already have [27 labels](https://github.com/qgis/QGIS/labels) that have being used (and created) for PR.

New labels: Bug, Feature request, Feedback, Priority: low, Priority: normal, Priority: high, Regression

Tracker:
Bug report → **Bug**
Feature request → **Feature request**

Status:
Open → NIL
In Progress → NIL
Feedback → **Feedback**
Reopened → NIL
Resolved → NIL
Closed → NIL
Rejected → NIL
Fixed for bounty → NIL
Bounty paid → NIL

Priority: 
Low → **Priority: low**
Normal → **Priority: normal**
High → **Priority: high**

Category: 
3D → NIL
Actions → NIL
Analysis library → NIL
...
Web Services clients/WMS → NIL
Web Services clients/XYZ → NIL
Windows Package → NIL
mac_os_specific → NIL

Affected QGIS version: (asked on the template)

Operating System:  (asked on the template)

Regression: → **Regression**

## Performance Implications

We really need help from GitHub team. It will take about 26 hours migration all issues, if we need to respected their rate limits. 

## Backwards Compatibility

All issue history 

## Votes

(required)

**Comment by haubourg** [comment]
@jgrocha I pushed a branch in my repo here with issue templates here : https://github.com/haubourg/QGIS/tree/add_issue_templates

**Comment by 3nids** [comment]
point 9. > put Redmine as read-only ?

**Comment by rduivenvoorde** [comment]
Well, it is not " difficult to maintain", because we run the vanilla debian packages now. It just runs :-)

**Comment by jgrocha** [comment]
> point 9. > put Redmine as read-only ?

Workflow updated.

**Comment by jgrocha** [comment]
> Well, it is not " difficult to maintain", because we run the vanilla debian packages now. It just runs :-)

I changed to "difficult to upgrade". Probably we should focus more on what we can get by moving to GitHub, instead of keep hitting the poor Redmine, that supported us for so many years.

**Comment by NathanW2** [comment]
>> Start mirroring QGIS repository on GitLab.

I'm ok with this generally we just have to make sure GitHub is the point of truth and people don't link or use the one at Gitlab

**Comment by gessel** [comment]
:(

**Comment by 3nids** [comment]
May I suggest to remove the low and medium priority levels? We would have only nothin or tagged as high priority.

**Comment by jidanni** [comment]
User sees:
>
Form
Fill your GitHub acount. **Do not provide the email associated with your GitHub account** (although you can log in using the account name or the email).

Name	￼
Email	￼
GitHub account	￼
￼Submit
If you don't have a GitHub account or if you prefer do not share it, you still be listed as the original author. The history of contributions will be preserved.

Which is nonsense. You need to say **Please provide the email associated with your GitHub account**

Only then does the form make sense.

**Comment by jef-n** [comment]
I thought gitlab was the final target - the mail doesn't mention this and in the QEP it also seems to be to github only - with a (remote?) possibility to move to gitlab later.

**Comment by jgrocha** [comment]
The first initial account we used to import all the issues was blocked by GitHub. We really need their help to do about 105k API calls without being blocked. The request was already sent.

We create another one to import a few issues so we can look at how they look like in GitHub. Please take a look at:
https://github.com/qgisissuebot/QGIS/issues

I think we aren't lost anything, but please confirm. Check open and closed issues. The link to the original issue is there, so you can put them side by side.

@jef-n Commit reference detection was improved. I think we were able to catch every reference.

@jef-n Links to specific comments/notes were preserved. There is one example https://github.com/qgisissuebot/QGIS/issues/129 that has the pointer to a note (right now the link still points to the Redmine issue; on a second pass, it will point to the comment on the corresponding GitHub issue).

@jef-n We also import the https://github.com/qgisissuebot/QGIS/issues/138 (the `proj_lpz_dist` crash) that has many references to others. The references to the other related issues were kept.

@elpaso You can see something like _Affected QGIS version: 3.4.0_ when it was provided in the original issue. This is just clear text on the GitHub issue. We can change it to label, or we can make it pointing to a version tag, but as it is, seems enough.

@3nids You proposal to remove ![](https://img.shields.io/static/v1.svg?label=&message=Priority:%20low&color=yellow) and ![](https://img.shields.io/static/v1.svg?label=&message=Priority:%20normal&color=yellow) seems adequately. The only really helpful is the Priority: high. More comments on this are welcome.

**Comment by jgrocha** [comment]
> 
> Only then does the form make sense.

@jidanni Thank you for your comment. I'm not sure if I fully understood your point. Please comment if I fail to answer properly.

If you provide us your email associated with your GitHub account, we can not use it to find your account. The API does not allow us to do that (they do not want people searching for accounts based on the email). 

This works:
```
curl https://api.github.com/users/jgrocha
```
This does not work:
```
curl https://api.github.com/users/jgr@di.uminho.pt
```

We ask for GitHub account to add a link to the user. If users do not provide the account, we simple don't link it.

Check the screen shot below from the previous import, were we mapped some users. The first user did provide a GitHub account. On mouse over event, the user is showed. For the second user we didn't have a GitHub account, so there is no link to him.

![image](https://user-images.githubusercontent.com/163681/54084519-ad328480-4329-11e9-8ade-6b2d08c94e6a.png)

**Comment by jgrocha** [comment]
We were blocked using the current GitHub API to upload all issues and comments to GitHub. I contact GitHub support and they are not willing to change the rate limits, as it _creates infrastructure problems with the number of notifications sent to repository watchers_. Not good news.

As an alternative, they pointed to a different API specific for this kind of migrations, that does not trigger messages to watchers. The documentation is available at: https://gist.github.com/jonmagic/5282384165e0f86ef105

I thought they were going to help us, but we will have to rewrite the importer. I'll report back after testing the alternative API.

**Comment by nyalldawson** [comment]
> :(

@gessel can you elaborate?

**Comment by gessel** [comment]
@nyalldawson, it is entirely off topic and irrelevant, but I like redmine, find it fairly easy to maintain, upgrade, and extend.  Philosophically it is for me in the same category as QGIS itself.  This doesn't matter much practically, but it has some meaning ideologically.  Migrating to a corporate, for profit, closed platform tends to contradict those values, for whatever that's worth.
Practically, I would not trust that Microsoft invested $7.5B for the good of the open source community and humanity at large; there was some projected return on investment equal to or greater than other options for that capital.  I don't yet see how that plays out. 
Gitlab seems more consistent, perhaps @jef-n has some thoughts. 
Overall, the success of the project is more important than symbolic gestures.  If Github is best for the project, I'm fully supportive.

**Comment by rduivenvoorde** [comment]
FYI: https://gitlab.com/qgis/QGIS/ is there to be used too (holding some old import/experiment I think; thanks @strk for 'claiming that one), maybe they are more friendly for imports like this one?

**Comment by nyalldawson** [comment]
@gessel

Thanks -- your feedback is certainly valued here!

The plan is actually to migrate from redmine -> github as a first step. Then, as a future piece of work it will be investigated whether a move from github to gitlab is both feasible and widely desired. But either way, getting the issues out of redmine and into the same place as the code repo is a ~~necessary~~ highly desirable step toward this.

**Comment by jidanni** [comment]
@jgrocha I see, at "github account" I am supposed to enter jidanni, not jidanni@jidanni.org.
Well OK then.

**Comment by uhliro26** [comment]
Priority changes from redmine were not migrated only comments in redmine. Is this per design?
See
https://issues.qgis.org/issues/21428#note-12
and
https://github.com/qgisissuebot/QGIS/issues/79#issuecomment-471195311

**Comment by haubourg** [comment]
> As an alternative, they pointed to a different API specific for this kind of migrations, that does not trigger messages to watchers. The documentation is available at: https://gist.github.com/jonmagic/5282384165e0f86ef105
> 
> I thought they were going to help us, but we will have to rewrite the importer. I'll report back after testing the alternative API.

Ouch, not a good move from Github's folks.

**Comment by jgrocha** [comment]
> The chronology of the
> priority changes from redmine were not migrated only comments in redmine. Is this per design?
> See
> https://issues.qgis.org/issues/21428#note-12 (2019-03-05-15-36)
> and
> [qgisissuebot/QGIS#79 (comment)](https://github.com/qgisissuebot/QGIS/issues/79#issuecomment-471195311)

Well spotted @uhliro26. The Priority change is a special comment with empty notes `{ "notes": "" }`. Maybe we can create a comment with some verbose message. I'll try and I'll report back.

**Comment by jgrocha** [comment]
**Great news!** Please take a look at https://github.com/qgib/QGIS/issues
All Redmine issues were imported (the Redmine export was on March, 7).

![Yabadabadoo](https://user-images.githubusercontent.com/163681/55189231-402a4600-5195-11e9-912c-c95bda0ca509.jpg)

Feel free to search and navigate on the issues. 

Issues were created by the bot account [qgib](https://github.com/qgib) (for _QGIS issue bot_). If you want to search for your issues, better to search with something like: `is:open mentions:jgrocha`

I also added @haubourg template proposal https://github.com/qgis/QGIS/pull/9438 for issues and for feature requests. You can try them on the bot account. You can report issues there ;-)

**Comment by 3nids** [comment]
@jgrocha awesome! How are handled next steps (read only on redmine, announcement, etc.)

**Comment by jgrocha** [comment]
# Migration plan (update)

The migration team needs the PSC engagement and its support to communicate with the large community of QGIS users and developers. We also need the help of the current QGIS infrastructure team to review the proposal and to provide the necessary authorizations (Redmine and Github) to run the process.

Teams:
- QGIS PSC ![](https://img.shields.io/static/v1.svg?label=&message=PSC&color=yellow)
- QGIS IT team ![](https://img.shields.io/static/v1.svg?label=&message=Infra&color=orange) Volunteers:
- QGIS Migration team ![](https://img.shields.io/static/v1.svg?label=&message=MigT&color=red). Volunteers: @jgrocha, @pdiasalmeida

## Preparation 

- [X] ![](https://img.shields.io/static/v1.svg?label=&message=MigT&color=red) Plan and develop the scripts 
- [X] ![](https://img.shields.io/static/v1.svg?label=&message=MigT&color=red) Get and prepare all maps (for versions, labels, statuses, categories, priorities, milestones and users) 
- [X] ![](https://img.shields.io/static/v1.svg?label=&message=MigT&color=red) Test the scripts using a full migration. All issues are available at [qgib/QGIS/issues](https://github.com/qgib/QGIS/issues)
- [ ]  ![](https://img.shields.io/static/v1.svg?label=&message=PSC&color=yellow) **Vote the proposal**
- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=PSC&color=yellow) **Decide the date**. We would like to have a window of 24 hours to do the migration, even if it take less than that 
- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=PSC&color=yellow) Use the media (Twitter, QGIS-User, QGIS-Developer) to announce the upcoming changes at the date defined

## Actual migration (on migration day)

- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=Infra&color=orange) Change Redmine to read only
- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=MigT&color=red) Export all issues from Redmine to individual json files
- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=Infra&color=orange) Grant permission to bot account [qgib](https://github.com/qgib/) on QGIS/qgis repository. Permission must be _admin_. This can be done with: `curl -X PUT -H "Content-Type: application/json" -d '{"permission":"admin"}' -H "Authorization: token <token>" https://api.github.com/repos/qgis/QGIS/collaborators/qgib` using a token from one admin account.
- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=MigT&color=red) Accept GitHub invitation by qgib
- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=Infra&color=orange) On Github, enable the “_Temporary interaction limits_” for a 24-hour period
- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=Infra&color=orange) On Github, enable issues on QGIS/qgis repository
- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=MigT&color=red) Run the issue migration to GitHub (first pass of two)
- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=MigT&color=red) Keep an eye on the logs (the script can break on the sever side)
- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=MigT&color=red) Run the issue migration to GitHub (second pass of two)
- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=MigT&color=red) Keep an eye on the logs (the script can break on the sever side)
- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=MigT&color=red) Check the logs and the results

## Post migration

- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=Infra&color=orange) On GitHub, merge pull request https://github.com/qgis/QGIS/pull/9438
- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=Infra&color=orange) On GitHub, merge pull request (upcoming PR to change the QGIS app dialogue regarding bug reporting)
- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=Infra&color=orange) On GitHub, disable the "Temporary interaction limits”
- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=Infra&color=orange) On GitHub, merge pull requests regarding documentation and web site (start by updating https://github.com/qgis/QGIS-Website/blob/master/source/site/getinvolved/development/bugreporting.rst)
- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=Infra&color=orange) Simple SQL script to update all issues on Redmine. Each line will be like: `insert into journals (journalized_id, journalized_type, user_id, notes, created_on) 
values (12345, 'Issue', 1122, 'Issue moved to https://github.com/qgis/QGIS/issues/98765', now() );`
- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=Infra&color=orange) On Redmine, update all issues on to point to their new location using the previous script
- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=PSC&color=yellow) Use the media (Twitter, QGIS-User, QGIS-Developer) to announce that the migration is over
- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=PSC&color=yellow) Post a new blog entry how to submit a issue
- [ ] ![](https://img.shields.io/static/v1.svg?label=&message=Infra&color=orange) Remove account qgib from QGIS repository, with `curl -I -X DELETE -H "Authorization: token <token>" https://api.github.com/repos/jgrocha/QGIS/collaborators/qgib` (or using the GitHub interface).

**Comment by nyalldawson** [comment]
Looks fantastic to me, great work

**Comment by jef-n** [comment]
I miss the categories and searches feel rather unpredicable to me.  Apparently I cannot search for exact phrases like `Redmine original issue: 2000` let along something like `category_id was changed from  .* to Processing/QGIS`.  Looks like github just splits them into pieces and searches for the individual words.
I cannot for instance find DXF only related bugs - it'll also dumps unrelated stuff, just because DXF appears in them in an otherwise unrelated note.   category_dxf however seems to work - and brings only the one issue I added it to.
BTW in ticket qgib/QGIS#17965 ([#20000](https://issues.qgis.org/issues/20000)) there is a forward reference to issue 20009 that is left as is.

**Comment by jgrocha** [comment]
@jef-n You are right. You found a bug (231 occurrences)! Thanks! If the issue number is followed by a carriage return, the script is not detecting it. On the input file, the carriage return is spelled literally, like `"notes": "Issue 1 is now #20009\r\n\r\nIssue 2 is related to #13997, ...`. I'll fix it and run it again against this repository.

**Comment by haubourg** [comment]
> I miss the categories

Yes, I do too. What about having a simplified label list and use a lookup table to affect the redmine issues to this list? 
I suggest we use prefixed naming for category labels. Something like `category:Labeling` for instance ?

**Comment by jgrocha** [comment]
We have more than 100 categories in Redmine. For new issues (created on GitHub) users will not use the "old" categories, but only the labels defined in GitHub.

For old issues, we can add a string for searching. We will use the string: 'category:label' (all lowercase, with '_' when there are spaces on the category).

**Comment by elpaso** [comment]
@jgrocha I think we are not really using 100 categories on redmine, probably the most used are around 20. But we definitely need categories, I use them a lot to restrict search.

**Comment by jgrocha** [comment]
> @jef-n You are right. You found a bug (231 occurrences)! Thanks! If the issue number is followed by a carriage return, the script is not detecting it. On the input file, the carriage return is spelled literally, like `"notes": "Issue 1 is now #20009\r\n\r\nIssue 2 is related to #13997, ...`. I'll fix it and run it again against this repository.

There was a problem in the second pass. It has been fixed. Thanks for pointing out @jef-n.

**Comment by jgrocha** [comment]
> @jgrocha I think we are not really using 100 categories on redmine, probably the most used are around 20. But we definitely need categories, I use them a lot to restrict search.

Agree! We will add the categories.

**Comment by 3nids** [comment]
about the migration day, is there a risk if someone tries to add an issue on github while the scripts are running? Have you tested it on qgib?

**Comment by haubourg** [comment]
@3nids I understood that this repo was for dry run testing. The real migration would imply a down time of issues.qgis.org. @jgrocha is that correct?

**Comment by jgrocha** [comment]
> @3nids I understood that this repo was for dry run testing. The real migration would imply a down time of issues.qgis.org. @jgrocha is that correct?

During the migration, on Github, we will enable the “Temporary interaction limits” for a 24-hour period. But this is not absolutely required, since we do all API requests synchronously. Personally, I prefer to prevent interaction for a day, as a precaution.

**Comment by jef-n** [comment]
Only tickets about QGIS should be copied (ie. not [#6220](https://issues.qgis.org/issues/6220) or [#16103](https://issues.qgis.org/issues/6220))

Searching for category works now - looking for categoryless tickets (`NOT category`) turned up these.

**Comment by jef-n** [comment]
Where's ticket [#21714](https://issues.qgis.org/issues/21714) or [#21683](https://issues.qgis.org/issues/21683) (looking for one that shows an image using `!foo.png!`).  Even looking for the pure number doesn't produce them - just unrelated tickets that happen to contain the number.

[#21379](https://issues.qgis.org/issues/21379) however has pictures (maybe because it's in a comment?).

**Comment by haubourg** [comment]
> I also added @haubourg template proposal [qgis/QGIS#9438](https://github.com/qgis/QGIS/pull/9438) for issues and for feature requests. You can try them on the bot account. You can report issues there ;-)

@jgrocha in fact, creating an issue does not propose issue templates on qgib repository, do I look into the right place?

**Comment by jef-n** [comment]
Can we modify the github css from within?  I currently use stylus with a style sheet to stop the waste of horizontal space.

**Comment by 3nids** [comment]
> about the migration day, is there a risk if someone tries to add an issue on github while the scripts are running? Have you tested it on qgib?

What I wanted to ask, is there any risk that any one not aware of the migration, just sees the bug tracker open on GH, creates a new issue and that it would cause troubles to the migration scripts?

**Comment by qgib** [comment]
> > I also added @haubourg template proposal [qgis/QGIS#9438](https://github.com/qgis/QGIS/pull/9438) for issues and for feature requests. You can try them on the bot account. You can report issues there ;-)
> 
> @jgrocha in fact, creating an issue does not propose issue templates on qgib repository, do I look into the right place?

@haubourg The templates can be tested at https://github.com/qgib/QGIS/issues/new/choose

I have to push them after each import.

Please notice that I've make a comment to yours PR https://github.com/qgis/QGIS/pull/9438. Templates must go to `.github/ISSUE_TEMPLATE` folder.

**Comment by qgib** [comment]
> > about the migration day, is there a risk if someone tries to add an issue on github while the scripts are running? Have you tested it on qgib?
> 
> What I wanted to ask, is there any risk that any one not aware of the migration, just sees the bug tracker open on GH, creates a new issue and that it would cause troubles to the migration scripts?

@3nids There is no problem for the script if someone creates a issue during the migration. Each call is synchronous. We submit an issue create call and wait for the number assigned to it before moving the the next one. The result is the map between old Redmine issues and new GitHub issues and the map is preserved even if there are other issues created during the migration.

**Comment by qgib** [comment]
All issues (up to March 7) were migrated (several times...). On the last exercise, we were able to do the migration without any break. The issue migration took about 15:30 hours.

The second run uses the issue map to fix references on descriptions and comments.

It is easy to see the second run edits. Just go to any issue and if there is an "edit" combobox on top of it, the issue was edited by the second run script. Example: https://github.com/qgib/QGIS/issues/19254#issuecomment-479334806

![issue edited](https://user-images.githubusercontent.com/48837220/55729008-344c4880-5a0d-11e9-8006-83a17de2f56c.png)

To know exactly what was changed on the second run, use that combobox  to show the most recent version of the description/comment.

![changes by the second run](https://user-images.githubusercontent.com/48837220/55729208-9c9b2a00-5a0d-11e9-89fb-8c300dd144da.png)

**Comment by uhliro26** [comment]
@jgrocha 
about chronology/history
the labeling is not in time with redmine and the comment in github
see:
comment
https://github.com/qgib/QGIS/issues/19354#issuecomment-479335727

labeling
https://github.com/qgib/QGIS/issues/19354#event-2248634170

**Comment by uhliro26** [comment]
@jgrocha
about Original Assignee / Current Assignee in the "issue-header" and assignees list
Originaly on opening the issue there was no assignee. So there was no "Original Assignee".

Shouldn't it not be something more correct to list the Original Assignee (if present) and the Current Assignee (which can change over time)?

Also in github currently there is no one assigned. shouldn't it alpaso
https://github.com/qgib/QGIS/issues/19354

**Comment by qgib** [comment]
> @jgrocha
> about Original Assignee / Current Assignee in the "issue-header" and assignees list
> Originaly on opening the issue there was no assignee. So there was no "Original Assignee".
> 
> Shouldn't it not be something more correct to list the Original Assignee (if present) and the Current Assignee (which can change over time)?
> 

You are right. We have all history of changes to the assignee in the comments (all history is there), but on the header we just put the "current" one. It would be better to call it "Current Assignee". Thank you for pointing it out.

> Also in github currently there is no one assigned. shouldn't it alpaso
> [qgib/QGIS#19354](https://github.com/qgib/QGIS/issues/19354)

We can't do it on this cloned repo. Only on the original QGIS repository it would be possible. You can only assign issues to project collaborators.

If you think it is important to assign issue to contributors, our proposal is to do it on the QGIS repository, after all issues are imported. We can do it later, by updating all issues based on the "current assignee" information, but only for GitHub QGIS contributors.

**Comment by qgib** [comment]
> @jgrocha
> about chronology/history
> the labeling is not in time with redmine and the comment in github
> see:
> comment 5.3.2019
> [qgib/QGIS#19354 (comment)](https://github.com/qgib/QGIS/issues/19354#issuecomment-479335727)
> 
> labeling 3.4.2019
> [qgib/QGIS#19354 (comment)](https://github.com/qgib/QGIS/issues/19354#event-2248634170)

You are right. We were able to preserve the dates and times when changes were made on Redmine. We can see exactly when the priority_id was changed from Normal to High, at 5.3.2019 as you mentioned. The history is preserved.

However, the Github labels are assigned at the time that they are created (when they are imported). The [API](https://gist.github.com/jonmagic/5282384165e0f86ef105) does not support it.

**Comment by mbernasocchi** [comment]
@jgrocha could you summarize the status. And please ping me whenever you think it is ready to be voted on by the PSC. 
@gioman did you have a look at the current result?

**Comment by jgrocha** [comment]
@mbernasocchi The current QEP proposal has been discussed and improved from the feedback of several developers. The script have been enhanced to incorporate the suggestions received. One of the enhancements added is to allow searches based the old Redmine categories. This can be done using search terms like: `category:symbology` or `category:vectors`.

We are been able to do all the migration at once without interruptions. We did it a few times and now the scripts are quite robust. This qgib account is _white flagged_ by the Github support, so we can do calls up to the API rate limit (they have other non documented filters in place that were blocking us. Only after talking with them we were able to reach the announced API rate limit.). We do it below the API rate limit. We need around 15 hours to do the whole migration.

We are preserving the full history of issues. All descriptions, comments and state changes are preserved. We are also mapping Redmine users to Github users, so you can do queries like: `mentions:mbernasocchi`.

To see the concrete result of the migration, we make it available to all at https://github.com/qgib/QGIS/issues. Everybody can see and check how the migration result looks like. We also published the templates prepared by @haubourg https://github.com/qgis/QGIS/pull/9438 for submitting new issues and feature requests. It easier to preview the templates.

### Limitations

We need a second run to update all references to issues, using the map between old Redmine issues and GitHub ones. On this second run, all occurrences of the regular expression `#[0-9]+` are replaced by the corresponding new issue number, as showed on https://github.com/qgis/QGIS-Enhancement-Proposals/issues/141#issuecomment-480840249.

But we found some false positives. Sometimes people refer to line numbers as `#24` or to colours like `#000077`. These occurrences are replaced by new issue ids, if they match and existing issue number. Users can check these changes and even revert any, if necessary. All changes are marked as edit (as explained in https://github.com/qgis/QGIS-Enhancement-Proposals/issues/141#issuecomment-480840249).

The most infamous issue where we have false positives is: https://github.com/qgib/QGIS/issues/17511. Even though, the replacement does not affect the readability of the issue.

### Next steps

We created a detailed plan for the migration https://github.com/qgis/QGIS-Enhancement-Proposals/issues/141#issuecomment-478296818

This plan also includes the update of Redmine, so every Redmine issue will have a pointer to its new instance at Github. People following old Redmine issue links on the email or on GIS StackExchange, for example, are able to go to current issue on Github.

Please take a look at the migration plan. We tried to cover every detail of the migration, including some tasks that we assigned to the PSC.

We presented the proposal in the last QGIS meeting and have it online for the last month. We also have the result of such migration available for debugging and for discussion. We sent a message (and a form) to all Redmine users. I think all are aware of this initiative and were able to contribute.

After one month of discussion and after integrating all improvements suggested by other developers, from my point of view, the proposal is ready to be reviewed and voted by the PSC.

**Comment by uhliro26** [comment]
@jgrocha and ATall others who are involved:
well done, well explained, hope I was not to nerving.
THX

**Comment by haubourg** [comment]
@jgrocha really nice work. Back to category topic, I fail to perform exact search over categories. I really think we still need a simplified category list in github labels.  Am I the only one?

**Comment by haubourg** [comment]
@jgrocha I tried a first attempt to reclassify categories:

I easily go down from 91 to 29 categories, but I think it is still a lot. Something like ~15/20 would be nicer. I discarded the category "documentation", assuming that this will move to the dedicated repository. 
Any idea to squeeze that down ? 

count | name | new label
-- | -- | --
1512 | GUI | GUI
1140 | Symbology | Symbology
963 | Unknown | 
935 | Vectors | Vectors
918 | Map Composer/Printing | Map Layout
777 | Build/Install | Build/Install
756 | Digitising | Digitising
557 | Processing/QGIS | Processing
519 | Processing/Core | Processing
518 | Python plugins | Plugins
499 | Data Provider | Data Provider
480 | Rasters | Raster
466 | GRASS | Processing
433 | Map Canvas | Map and legend
402 | DB Manager | Database
371 | Projection Support | Projection Support
351 | Web Services clients/WMS | Data Provider
340 | Project Loading/Saving |  Project
329 | Labelling | Labeling
321 | C++ Plugins | Plugins
311 | Map Legend | Map and legend
282 | QGIS Server | QGIS Server
282 | Attribute table | Attribute table
227 | Data Provider/PostGIS | Data Provider
199 | Processing/Modeller | Processing
195 | Browser |  
175 | Web Services clients/WFS | Data Provider
173 | Processing/GDAL | Processing
163 | Processing/GRASS | Processing
163 | Forms | Forms
151 | Documentation and Help | 
150 | Expressions | Expressions
148 | Data Provider/OGR | Data Provider
140 | GDAL Tools | Raster
126 | Processing/SAGA | Processing
120 | Edit widget | Forms
104 | Data Provider/MSSQL | Data Provider
101 | Processing/GUI | Processing
96 | Editing | Editing
87 | Map Tools | Map tools
82 | Data Provider/Oracle | Data Provider
82 | Translations and international | Translations and international
80 | Plugin Manager | Plugins
79 | Windows Package | Build/Install
78 | C++ plugins/Georeferencer | Plugins
77 | 3D | 3D
77 | PyQGIS Console | PyQGIS
67 | Actions | Actions
66 | Data Provider/SpatiaLite | Data Provider
57 | Field calculator | Expressions
54 | Python bindings / sipify | PyQGIS
51 | Geometry | Expressions
44 | mac_os_specific | Build/Install
40 | MetaSearch Catalogue Client | Plugins
39 | DXF export | CAD
37 | Analysis library | API
37 | Data Provider/Delimited Text | Data Provider
36 | Diagrams | Symbology
34 | Relations | Forms
34 | OsX UI | Build/Install
33 | Processing/OGR | Processing
30 | C++ plugins/Geometry Checker | Plugins
29 | Data Source Manager | Data Provider
29 | Customisation Framework | GUI
27 | Raster Calculator | Raster
27 | C++ plugins/Offline editing | Plugins
26 | Authentication system | Authentication
25 | Tests Suite | Quality Assessment
25 | Virtual Layers | Database
23 | C++ plugins/Topology checker | Plugins
21 | Data Provider/MDAL | Data Provider
21 | C++ plugins/Globe | Plugins
20 | Web Services clients/WCS | Data Provider
20 | C++ plugins/GPS plugin | Plugins
18 | Decorations | Map and legend
17 | Network | API
17 | DWG/DXF import | CAD
16 | Web Services clients/XYZ | Data Provider
15 | Web Services clients/ArcGIS | Data Provider
13 | Virtual Fields | Database
11 | Metadata | Metadata
9 | C++ plugins/Evis | Plugins
8 | GPS Live Tracking | GPS
8 | C++ plugins/Spatial Query | Plugins
8 | Simplification | Expressions
4 | Data Provider/GeoNode | Data Provider
3 | Processing/OTB | Processing
2 | Overview | Map and legend
2 | C++ plugins/Zonal statistics | Plugins
1 | OpenCL |  
1 | Processing/WPS | Processing

**Comment by jgrocha** [comment]
@haubourg Looks great to me. 29 labels are not too much. If we look around (other **successful** Github projects) we can see many with 100 and more labels.

Two comments:
1. I would keep the `Map Composer/Printing` as it is right now. `Map Layout` sound too general (and has more interception with `Map and legend`).
2. `GRASS` can be under `Processing`, but we have so many issues related with GRASS, that I would keep `GRASS` as a label, so our GRASS experts (and people from GRASS team) can pick them quickly.

**Comment by nyalldawson** [comment]
I'd suggest "print layouts" is a more modern term for the composer label. I wonder about "Actions". We could merge "C++ plugins" and "plugins" (most users wouldn't know the difference anyway!). Quality assessment could also just use the existing "co/infrastructure" label? That drops another 3 at least 😀

**Comment by gioman** [comment]
One minor suggestion: can we have different colors (not red, maybe orange/yellow) for "regression" and "high priority"? Also I think we miss a label for "causes crash or data corruption" (?), if confirmed I think that this is an important flag of Redmine tickets we should keep here on github.

**Comment by qgib** [comment]
> One minor suggestion: can we have different colors (not red, maybe orange/yellow) for "regression" and "high priority"? Also I think we miss a label for "causes crash or data corruption" (?), if confirmed I think that this is an important flag of Redmine tickets we should keep here on github.

1) Yes, using a different colour might be better.

2) Good point! Can we call this label just "Data corruption"? I'll improve the migration script to catch this on old Redmine issues.

**Comment by gioman** [comment]
>     2\. Good point! Can we call this label just "Data corruption"? I'll improve the migration script to catch this on old Redmine issues.

the "crash" label is more important (is fundamental), if we have to choose, but if can have both is better.

**Comment by haubourg** [comment]
> One minor suggestion: can we have different colors (not red, maybe orange/yellow) for "regression" and "high priority"? Also I think we miss a label for "causes crash or data corruption" (?), if confirmed I think that this is an important flag of Redmine tickets we should keep here on githu

I totally agree, the color code we choose is almost as important as the text itself. 
Let's try to put some rules so that we keep some readability when searching for issues with many label tags, with the help of http://colorbrewer2.org :

- Primary red only for regressions #f00

- crash label could be black

![image](https://user-images.githubusercontent.com/1629853/56115026-c2bb4f80-5f62-11e9-9b91-43c4cc23531f.png)


- Priorities should be in a coherent color ramp. I'd suggest this [one](http://colorbrewer2.org/#type=sequential&scheme=OrRd&n=3)

![image](https://user-images.githubusercontent.com/1629853/56114870-62c4a900-5f62-11e9-9bb5-d1fd1c8083b3.png)

- All other categories should be using soft colors, grouped by family if we have some kind of groups. I'll try to see if this is possible.

**Comment by jgrocha** [comment]
![migration-issues-to-github](https://user-images.githubusercontent.com/163681/56474920-b8c0b180-6478-11e9-8b0f-64b7ba483a22.png)

We have done another successful migration, adding much more Github labels, according to the suggestions of @haubourg and @gioman. We have changed a few colours (but this can be done later, at any time). The Redmine snapshot is from 2019-04-13.

Please check the result at: https://github.com/qgib/QGIS/issues

**Bug** and **Feature request** templates are enabled for testing on this _qgib_ repository.

If someone wants to have access to this https://github.com/qgib/ repository, feel free to ask for the credentials (for organizing issues with projects boards, to play with label colour, etc).

Although not very important, we have fixed the majority of false positives references to issues from thread dumps, where each line starts with `#1`, `#2`, etc.  Example: https://github.com/qgib/QGIS/issues/17309 (only references to issues were made).

**Comment by mbernasocchi** [comment]
@jgrocha this all looks great to me. do you want me to trigger a psc vote today?

**Comment by qgib** [comment]
Yes, yes, please! Everything is ready! Thank you! -+-+-+-+-+-+-+-+-+-Geomaster, LdaRua Afonso Palmeira, 31, 7 D4715-278 BragaNIF 510906109Tm +351 910333888Email jgr@geomaster.ptSent from my mobile. Please excuse the brevity, spelling and punctuation. 
-------- Mensagem original --------De : Marco Bernasocchi <notifications@github.com> Data: 06/05/19  18:51  (GMT+00:00) Para: qgis/QGIS-Enhancement-Proposals <QGIS-Enhancement-Proposals@noreply.github.com> Cc: QGIS issue bot <qgib@geomaster.pt>, Comment <comment@noreply.github.com> Assunto: Re: [qgis/QGIS-Enhancement-Proposals] Migrating issues from Redmine
  to GitHub (#141) @jgrocha this all looks great to me. do you want me to trigger a psc vote today?

—You are receiving this because you commented.Reply to this email directly, view it on GitHub, or mute the thread.
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/qgis/QGIS-Enhancement-Proposals/issues/141#issuecomment-489711992",
"url": "https://github.com/qgis/QGIS-Enhancement-Proposals/issues/141#issuecomment-489711992",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by luipir** [comment]
congratulation to all worked in this and in this GEP :)

**Comment by haubourg** [comment]
Yeah! Kudos @jgrocha for tackling a  such long running issue :+1:

**Comment by mbernasocchi** [comment]
just for reference here the:
- [go-ahead message]( https://lists.osgeo.org/pipermail/qgis-developer/2019-May/057388.html)
- [Planed schedule](https://lists.osgeo.org/pipermail/qgis-developer/2019-May/057383.html)

**Comment by qgib** [comment]
The migration itself has ended and the interaction limitation was removed. Issue and feature request templates are in place and workig as expected.

We still need to do some post processing. Work is not over yet.

Tomorrow I'll report back here.

**Comment by qgib** [comment]
@jef-n I've fixed the markdown translation of inline images for issues still open.

**Comment by jef-n** [comment]
@jgrocha Thanks.   Example: qgis/QGIS#29945

BTW I just closed qgis/QGIS#29862 - I knew that there was a ticket, but I couldn't find it here by searching for "widget.value" - redmine did find it - turns out looking for "self.widget.value" also does find it here.  Any idea why that is?

**Comment by jef-n** [comment]
> BTW I just closed [qgis/QGIS#29862](https://github.com/qgis/QGIS/issues/29862) - I knew that there was a ticket, but I couldn't find it here by searching for "widget.value" - redmine did find it - turns out looking for "self.widget.value" also does find it here. Any idea why that is?

Apparently you cannot search for substrings in github

**Comment by qgib** [comment]
@jef-n @rduivenvoorde 

@strk already asked for the map from Redmine to GitHub. Can one of you review the following SQL code and run it against our redmine database?

Please do the `insert` for testing and chech if it displays properly. If it does, you can run the loop. Feel free to change and improved it.

I can do it, if you provide me the postgresql connect string (with the credentials).

The map file to load into the database is attached.

```SQL
-- create table with map
CREATE TABLE redmine2github_map (redmine_id integer, github_id integer);
COPY redmine2github_map FROM '/home/jgr/bonn/ownCloud/geomaster/QGisUserConf 2019/map.csv' csv header;

--test with one
insert into journals (journalized_id, journalized_type, user_id, notes, created_on) values (10245, 'Issue', 1122, 'Issue moved to https://github.com/qgis/QGIS/issues/18682', now() );
-- check the output at: https://issues.qgis.org/issues/10245

-- write a loop to update every issue
DO $do$
DECLARE
    pair RECORD;
BEGIN
    FOR pair in SELECT redmine_id, github_id FROM redmine2github_map limit 10
    loop
    	raise notice 'Redmine issue %', format('Issue %s moved to %s', pair.redmine_id, pair.github_id);	
    	EXECUTE format($$ insert into journals (journalized_id, journalized_type, user_id, notes, created_on) values (%s, 'Issue', 1122, 'Issue moved to https://github.com/qgis/QGIS/issues/%s', now() ) $$, pair.redmine_id, pair.github_id);
    END LOOP;
END;
$do$;
```
[map.zip](https://github.com/qgis/QGIS-Enhancement-Proposals/files/3220410/map.zip)

**Comment by jef-n** [comment]
> The map file to load into the database is attached.

I added a custom fields and filled it with the map instead ("Copied to github as #").

**Comment by jgrocha** [comment]
The technical part of the migration is over. All issues are on GitHub and the templates for reporting issues are in place. 

We have a bidirectional map between Github and Redmine (thanks to @jef-n), so it is easy to switch between them. It really nice to be able to move between the two platforms with just one click.

![image](https://user-images.githubusercontent.com/163681/58384051-442be600-7fd5-11e9-8f47-1c12d96a50d3.png)

We need to update and improve the documentation, use QGIS blog to tell users about the changes, update our workshop materials, and so on. It will take some time.

I've already sent the bot credentials (qgib account) to the PSC and I'll no longer use that account.

We still have a problem… more than 3000 open issues to manage! QGIS success creates a whole set of problems, right?

Let's start cleaning the issue queue, but first we need to express our gratitude for all the valuable work of those supporting Redmine until today. There were 19844 issues and that means a lot of triage effort and maintenance behind the scenes. Please join me in saying "Thank You" to @gioman, @jef-n, @rduivenvoorde and others running Redmine (since 2010).

**Comment by gioman** [comment]
@jgrocha and all others involved, thanks for the effort

> We still have a problem… more than 3000 open issues to manage! QGIS success creates a whole set of problems, right?

They are not more than they were on Redmine :) Actually the real bug are "only"  1283, and of those ~250 are awaiting feedback for more than 15 days, so likely to be closed as not reproducible or because already fixed. Usually those tickets (bugs, in feedback since long) are the ones I focus when I start some cleanup.

Possibly a large chunk of feature requests could be closed (as already implemented) but I feel they don't get much attention (personally I don't look to feature requests as frequently as I do for bugs).


> There were 19844 issues and that means a lot of triage effort and maintenance behind the scenes. Please join me in saying "Thank You" to @gioman, @jef-n, @rduivenvoorde and others running Redmine (since 2010).

Thank you, after 1084 tickets open and 24053 comments (both high scores) on Redmine let's see how it plays out doing the same here.

**Comment by strk** [comment]
For reference, I've filed a ticket for Gitea to request allowing commit-metadata based redirection of issue references: https://github.com/go-gitea/gitea/issues/7084 -- not sure it'll ever be implemented

**Comment by jef-n** [comment]
@jgrocha issue qgis/QGIS#29494 references ticket #21367, which actually is the redmine ticket number.  It should reference qgis/QGIS#29184 instead.

**Comment by jgrocha** [comment]
> @jgrocha issue [qgis/QGIS#29494](https://github.com/qgis/QGIS/issues/29494) references ticket #21367, which actually is the redmine ticket number. It should reference [qgis/QGIS#29184](https://github.com/qgis/QGIS/issues/29184) instead.

I can't find an explanation. The issue number was detected. In the the migration log we can see that two issues were identified, as expected. The mapping pair is present in the mapping.

```text
May 25 21:30:07 (19393) Loading redmine issue: [21678] from file [21678.json]
May 25 21:30:07 ----------------------------------------
May 25 21:30:07 29488
May 25 21:30:07 ----------------------------------------
May 25 21:30:07 Match (2nd) found: 21367
May 25 21:30:07 ----------------------------------------
May 25 21:30:07 ----------------------------------------
May 25 21:30:07 Match (2nd) found: 21672
May 25 21:30:07 ----------------------------------------
May 25 21:30:07 Comment updated!
May 25 21:30:07 Redmine related issues: [21672](https://issues.qgis.org/issues/21672)
May 25 21:30:07 Github related issues: #29488 (duplicates)
May 25 21:30:07 Redmine sub-issues:
May 25 21:30:07 Github sub-issues:
May 25 21:30:07 Issue updated!
```
There is no error reported from the GitHub side.

I've identified similar issues (with the same structure) and all were rewritten properly. If we can find another one where the ID was not replaced, maybe we can understand where the script failed.

If I find any clue, I'll return to you.

**Comment by jef-n** [comment]
@jgrocha the updated description from [github_issue_maker.py:246](https://gitlab.com/jgrocha/redmine2github/blob/master/src/github_issues/github_issue_maker.py#L246) is apparently not used anywhere

**Comment by jgrocha** [comment]
> @jgrocha the updated description from [github_issue_maker.py:246](https://gitlab.com/jgrocha/redmine2github/blob/master/src/github_issues/github_issue_maker.py#L246) is overwritten when there are related tickets. Does https://gist.github.com/jef-n/f6f0f52d577e615bfcc6dc5ae2b66468 help?

@jef-n Thank god you exist! The problem are the ones with issues references on the description and related issues, as you spotted. 

Doing a query (on Redmine!), it reports 222 issues, but some of them are false positives. _I've done a query searching for bugs or feature requests, having related issues and `#` on the description._ The ones with references to issues are less than that.

My proposal is check each one and do the proper replacement. It is dangerous to run the migration script again, because we might change issues already mapped. I've wrote a small script to do this changes, and it only changes the part of the description. (The issue description is divided in tree parts and we can trust the `\n---\n` separator).

My proposal is to call it for each issue (from the `github_candidates.csv` file), like:

```
$ python3 issues.py 29637
Redmine issue not in map
Redmine issue not in map
Redmine issue not in map
Number of replacements: 0
Skip issue 29637
$ python3 issues.py 29598
Redmine 20921 mapped to 28740
Redmine 20980 mapped to 28799
Redmine 21648 mapped to 29464
Number of replacements: 3
```
and so on.

The script and the two supporting files are at: https://gist.github.com/jgrocha/be7853392bc921f4be23d4e704f933e5

@jef-n Can you take a look at the script?
https://gist.github.com/jgrocha/be7853392bc921f4be23d4e704f933e5#file-issues-py

Right now, for testing, the script creates a copy of the issue in one of my repositories. By changing the authentication and uncomment the `issue.edit(body=newbody)` we get it running on our qgis/QGIS repo. Doing one by one is safe, I think, and I can revert any if something goes wrong.

**Comment by jgrocha** [comment]
I've fixed the ids that weren't mapped at the migration time. 153 issues were edited and 181 references were fixed. Thanks @jef-n for spotting the missing mappings.
